rcl_action/src/rcl_action/action_client.c

```diff
@@ -0,0 +1,723 @@
+
+  *feedback_subscription = rcl_get_zero_initialized_subscription();
+
+  ret = rcl_subscription_init(
+    feedback_subscription, node,
+    feedback_message_type_support,
+    feedback_topic_name,
+    feedback_subscription_options);
+
+  allocator.deallocate(feedback_topic_name, allocator.state);
+
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_SERVICE_NAME_INVALID == ret) {
+      ret = RCL_RET_ACTION_NAME_INVALID;
+    } else if (RCL_RET_BAD_ALLOC != ret) {
+      ret = RCL_RET_ERROR;
+    }
+  }
+  return ret;
+}
+
+static rcl_ret_t rcl_action_status_subscription_init(
+  rcl_subscription_t * status_subscription, const rcl_node_t * node,
+  const rosidl_message_type_support_t * status_message_type_support,
+  const char * action_name,
+  const rcl_subscription_options_t * status_subscription_options)
+{
+  assert(NULL != status_subscription);
+  assert(NULL != node);
+  assert(NULL != status_message_type_support);
+  assert(NULL != action_name);
+  assert(NULL != status_subscription_options);
+
+  rcl_ret_t ret;
+  rcl_allocator_t allocator = status_subscription_options->allocator;
+
+  char * status_topic_name = NULL;
+  ret = rcl_action_get_status_topic_name(
+    action_name, allocator, &status_topic_name);
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_BAD_ALLOC != ret) {
+      ret = RCL_RET_ERROR;
+    }
+    return ret;
+  }
+
+  *status_subscription = rcl_get_zero_initialized_subscription();
+
+  ret = rcl_subscription_init(
+    status_subscription, node,
+    status_message_type_support,
+    status_topic_name,
+    status_subscription_options);
+
+  allocator.deallocate(status_topic_name, allocator.state);
+
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_SERVICE_NAME_INVALID == ret) {
+      ret = RCL_RET_ACTION_NAME_INVALID;
+    } else if (RCL_RET_BAD_ALLOC != ret) {
+      ret = RCL_RET_ERROR;
+    }
+  }
+  return ret;
+}
+
+rcl_ret_t
+rcl_action_client_init(
+  rcl_action_client_t * action_client,
+  rcl_node_t * node,
+  const rosidl_action_type_support_t * type_support,
+  const char * action_name,
+  const rcl_action_client_options_t * options)
+{
+  RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Initializing action client");
+  RCL_CHECK_ARGUMENT_FOR_NULL(options, RCL_RET_INVALID_ARGUMENT);
+  rcl_allocator_t * allocator = (rcl_allocator_t *)&options->allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(
+    allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(action_client, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(node, RCL_RET_INVALID_ARGUMENT);
+  if (!rcl_node_is_valid(node)) {
+    return RCL_RET_NODE_INVALID;
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(type_support, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(action_name, RCL_RET_INVALID_ARGUMENT);
+  RCUTILS_LOG_DEBUG_NAMED(
+    ROS_PACKAGE_NAME, "Initializing client for action name '%s'", action_name);
+  // Allocate space for the implementation struct.
+  rcl_action_client_impl_t *impl = allocator->allocate(
```

Comment:
Should error if `NULL != action_client->impl` before allocating to avoid leaking memory if `action_client` is already initialized.

Reply:
Good catch. Done in c5dabd05bc66f9c01431ec5cfe1b86815081c823.